### PR TITLE
Fix ORDER BY on aggregate

### LINKED ISSUE #1479: ORDER BY min(x) results in error  `Plan("No field named 'foo.x'. Valid fields are 'MIN(foo.x)'.")`
**Describe the bug**
You can not refer to an aggregate directly in the ORDER BY clause

**To Reproduce**
```shell
echo "1" > /tmp/foo.csv
cargo run datafusion-cli
```

```sql
DataFusion CLI v5.1.0

❯ create external table foo(x int) stored as CSV location '/tmp/foo.csv';
0 rows in set. Query took 0.033 seconds.

❯ select min(x) from foo order by min(x);
Plan("No field named 'foo.x'. Valid fields are 'MIN(foo.x)'.")
```

**Expected behavior**
The query should produce a single column, with a single row `1`

Note there are several workarounds to express this query that do work:

Refer to the item by index:

```sql
❯ select min(x) from foo order by 1;
+------------+
| MIN(foo.x) |
+------------+
| 1          |
+------------+
1 row in set. Query took 0.045 seconds.
```

Given the query an alias:
```sql
❯ select min(x) as min_x from foo order by min_x;
+-------+
| min_x |
+-------+
| 1     |
+-------+
1 row in set. Query took 0.010 seconds.
```

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1479.

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/src/logical_plan/expr.rs [hunk 1]
```diff
@@ -1317,6 +1319,83 @@ pub fn normalize_cols(
...
+        .collect()
+}
+
+fn rewrite_sort_col_by_aggs(expr: Expr, plan: &LogicalPlan) -> Result<Expr> {
+    match plan {
+        LogicalPlan::Aggregate(Aggregate {
+            input,
+            group_expr: _,
+            aggr_expr,
+            schema: _,
```
Comment (Jimexist) [inline datafusion/src/logical_plan/expr.rs:1357]: i guess you can use `..` to ignore these fields?

File: datafusion/src/logical_plan/expr.rs [hunk 2]
```diff
@@ -1317,6 +1319,83 @@ pub fn normalize_cols(
         .collect()
 }
 
+/// Rewrite sort on aggregate expressions to sort on the column of aggregate output
+#[inline]
```
Comment (Jimexist) [inline datafusion/src/logical_plan/expr.rs:1323]: why would this need to be inlined? or can't the compiler infer that?

Comment (viirya) [reply datafusion/src/logical_plan/expr.rs:1323]: Hmm, I saw there is `#[inline]` for `normalize_cols` above. `rewrite_sort_cols_by_aggs`'s looks similar (i.e., `into_iter().map().collect()`), so I added it too here. Should I remove it?

File: datafusion/src/logical_plan/expr.rs (hunk 1 above)
Comment (viirya) [reply datafusion/src/logical_plan/expr.rs:1357]: yea. i will use it.

File: datafusion/src/logical_plan/expr.rs (hunk 2 above)
Comment (xudong963) [reply datafusion/src/logical_plan/expr.rs:1323]: > In Rust, a unit of (separate) compilation is a crate. If a function f is defined in a crate A, then all calls to f from within A can be inlined, as the compiler has full access to f. If, however, f is called from some downstream crate B, such calls can’t be inlined. B has access only to the signature of f, not its body.

So I think both of them don't need `#[inline]`.

Comment (viirya) [reply datafusion/src/logical_plan/expr.rs]: Ok, removed.

File: datafusion/src/logical_plan/expr.rs [hunk 3]
```diff
@@ -1317,6 +1318,79 @@ pub fn normalize_cols(
...
+                    nulls_first,
+                } => {
+                    let sort = Expr::Sort {
+                        expr: Box::new(rewrite_sort_col_by_aggs(*expr, plan)?),
+                        asc,
+                        nulls_first,
+                    };
+                    Ok(sort)
+                }
+                _ => Ok(expr),
```
Comment (alamb) [inline datafusion/src/logical_plan/expr.rs:1343]: I think you can avoid this clone via:

```suggestion
            match expr {
                Expr::Sort {
                    expr,
                    asc,
                    nulls_first,
                } => {
                    let sort = Expr::Sort {
                        expr: Box::new(rewrite_sort_col_by_aggs(*expr, plan)?),
                        asc,
                        nulls_first,
                    };
                    Ok(sort)
                }
                expr => Ok(expr),
```

File: datafusion/src/logical_plan/expr.rs [hunk 4]
```diff
@@ -1317,6 +1318,79 @@ pub fn normalize_cols(
         .collect()
 }
 
+/// Rewrite sort on aggregate expressions to sort on the column of aggregate output
```
Comment (alamb) [inline datafusion/src/logical_plan/expr.rs:1321]:
```suggestion
/// Rewrite sort on aggregate expressions to sort on the column of aggregate output
/// For example, `max(x)` is written to `col("MAX(x)")` 
```

It took me a bit of thought to understand what this was doing so I think an extra comment might help.

File: datafusion/tests/sql/order.rs [hunk 5]
```diff
@@ -32,6 +32,27 @@ async fn test_sort_unprojected_col() -> Result<()> {
...
+    let expected = vec![
+        "+-----------------------------+",
+        "| MIN(aggregate_test_100.c12) |",
+        "+-----------------------------+",
+        "| 0.01479305307777301         |",
+        "+-----------------------------+",
+    ];
+    assert_batches_eq!(expected, &actual);
+
+    let sql = "SELECT MIN(c12) FROM aggregate_test_100 ORDER BY MIN(c12) + 0.1";
```
Comment (alamb) [inline datafusion/tests/sql/order.rs:50]: 👍

File: datafusion/src/logical_plan/expr.rs [hunk 6]
```diff
@@ -1317,6 +1318,79 @@ pub fn normalize_cols(
...
+                    if found_agg.is_some() {
+                        let agg = normalize_col(found_agg.unwrap().clone(), self.plan)?;
+                        let col = Expr::Column(
+                            agg.to_field(self.input.schema())
+                                .map(|f| f.qualified_column())?,
+                        );
+                        Ok(col)
+                    } else {
+                        Ok(expr)
+                    }
```
Comment (alamb) [inline datafusion/src/logical_plan/expr.rs:1379]:
```suggestion
                    if let Some(found_agg) = self.aggr_expr.iter().find(|a| (**a) == normalized_expr) {
                        let agg = normalize_col(found_agg, self.plan)?;
                        let col = Expr::Column(
                            agg.to_field(self.input.schema())
                                .map(|f| f.qualified_column())?,
                        );
                        Ok(col)
                    } else {
                        Ok(expr)
                    }
```

Comment (alamb) [reply datafusion/src/logical_plan/expr.rs:1379]: I think you can avoid some `is_some()` and `unwrap()` calls here to make the code stylistically nicer

**Review by alamb — APPROVED** [review]
This looks great. Thank you @viirya  🏅 

I also played around with it locally and it worked like a charm 👍 . 

I had some stylistic suggestions but I also think this PR could be merged as is as well.

**Comment by viirya** [comment]
Thank you @alamb . Committed the suggestions.

**Comment by alamb** [comment]
Thanks again @viirya !

**Comment by viirya** [comment]
Thank you @alamb @Jimexist @xudong963 !

**Pull request merged**